File: src-cpp/DesktopFlyouts.Shared/DesktopFlyoutIsland.cpp
#include "pch.h"
#include "DesktopFlyoutIsland.h"
#include "DesktopFlyoutIslandTemplateSettings.h"
#include "DesktopFlyout.h"
#if __has_include("DesktopFlyoutIsland.g.cpp")
#include "DesktopFlyoutIsland.g.cpp"
#endif

namespace winrt::DesktopFlyouts::implementation
{
    winrt::Microsoft::UI::Xaml::DependencyProperty DesktopFlyoutIsland::s_islandWidthProperty =
        winrt::Microsoft::UI::Xaml::DependencyProperty::Register(
            L"IslandWidth",
            winrt::xaml_typename<winrt::Microsoft::UI::Xaml::GridLength>(),
            winrt::xaml_typename<winrt::DesktopFlyouts::DesktopFlyoutIsland>(),
            winrt::Microsoft::UI::Xaml::PropertyMetadata{ winrt::box_value(winrt::Microsoft::UI::Xaml::GridLengthHelper::Auto()) });

    winrt::Microsoft::UI::Xaml::DependencyProperty DesktopFlyoutIsland::s_islandHeightProperty =
        winrt::Microsoft::UI::Xaml::DependencyProperty::Register(
            L"IslandHeight",
            winrt::xaml_typename<winrt::Microsoft::UI::Xaml::GridLength>(),
            winrt::xaml_typename<winrt::DesktopFlyouts::DesktopFlyoutIsland>(),
            winrt::Microsoft::UI::Xaml::PropertyMetadata{ winrt::box_value(winrt::Microsoft::UI::Xaml::GridLengthHelper::Auto()) });

    DesktopFlyoutIsland::DesktopFlyoutIsland() :
        m_templateSettings(winrt::make<implementation::DesktopFlyoutIslandTemplateSettings>())
    {
        DefaultStyleKey(winrt::box_value(L"DesktopFlyouts.DesktopFlyoutIsland"));
        m_cornerRadiusToken = RegisterPropertyChangedCallback(
            winrt::Microsoft::UI::Xaml::Controls::Control::CornerRadiusProperty(),
            [this](auto const&, auto const&) { UpdateTemplateSettings(); });
        UpdateTemplateSettings();
    }

    DesktopFlyoutIsland::~DesktopFlyoutIsland()
    {
        if (m_cornerRadiusToken != 0)
        {
            UnregisterPropertyChangedCallback(winrt::Microsoft::UI::Xaml::Controls::Control::CornerRadiusProperty(), m_cornerRadiusToken);
        }
    }

    winrt::Microsoft::UI::Xaml::DependencyProperty DesktopFlyoutIsland::IslandWidthProperty()
    {
        return s_islandWidthProperty;
    }

    winrt::Microsoft::UI::Xaml::DependencyProperty DesktopFlyoutIsland::IslandHeightProperty()
    {
        return s_islandHeightProperty;
    }

    winrt::Microsoft::UI::Xaml::GridLength DesktopFlyoutIsland::IslandWidth()
    {
        return winrt::unbox_value<winrt::Microsoft::UI::Xaml::GridLength>(GetValue(s_islandWidthProperty));
    }

    void DesktopFlyoutIsland::IslandWidth(winrt::Microsoft::UI::Xaml::GridLength const& value)
    {
        SetValue(s_islandWidthProperty, winrt::box_value(value));
    }

    winrt::Microsoft::UI::Xaml::GridLength DesktopFlyoutIsland::IslandHeight()
    {
        return winrt::unbox_value<winrt::Microsoft::UI::Xaml::GridLength>(GetValue(s_islandHeightProperty));
    }

    void DesktopFlyoutIsland::IslandHeight(winrt::Microsoft::UI::Xaml::GridLength const& value)
    {
        SetValue(s_islandHeightProperty, winrt::box_value(value));
    }

    winrt::DesktopFlyouts::DesktopFlyoutIslandTemplateSettings DesktopFlyoutIsland::TemplateSettings()
    {
        return m_templateSettings;
    }

    void DesktopFlyoutIsland::UpdateTemplateSettings()
    {
        auto radius = CornerRadius();
        auto inner = [](double value) { return std::max(0.0, value - 1.0); };
        m_templateSettings.SetValue(
            winrt::DesktopFlyouts::DesktopFlyoutIslandTemplateSettings::BackdropCornerRadiusProperty(),
            winrt::box_value(winrt::Microsoft::UI::Xaml::CornerRadius{
                inner(radius.TopLeft), inner(radius.TopRight), inner(radius.BottomRight), inner(radius.BottomLeft) }));
    }
}

File: src/DesktopFlyouts.Shared/DesktopFlyoutIsland.cs
// Copyright (c) 0x5BFA. All rights reserved.
// Licensed under the MIT license.

using System;
using Windows.Foundation;

#if UWP
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
#elif WASDK
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
#endif

namespace DesktopFlyouts
{
    /// <summary>
    /// Represents a content section inside a <see cref="DesktopFlyout"/>.
    /// </summary>
    /// <remarks>
    /// Islands are arranged by their owner flyout according to
    /// <see cref="DesktopFlyout.IslandsOrientation"/>. Put the XAML content for one visual section
    /// in each island.
    /// </remarks>
#if WASDK && !HAS_UNO
    [WinRT.GeneratedBindableCustomProperty([nameof(TemplateSettings)], [])]
#endif
    public partial class DesktopFlyoutIsland : ContentControl
    {
        private const double LayoutEpsilon = 0.5D;

        private WeakReference<DesktopFlyout>? _owner;
        private long _propertyChangedCallbackTokenForCornerRadiusProperty;
        private Size _lastActualSize;
        private Size _lastDesiredSize;
        private bool _hasLayoutSnapshot;

        /// <summary>
        /// Gets an object that provides calculated values that can be referenced from the island template.
        /// </summary>
        /// <value>The calculated template settings for this island.</value>
        public DesktopFlyoutIslandTemplateSettings TemplateSettings { get; } = new();

        /// <summary>
        /// Initializes a new instance of <see cref="DesktopFlyoutIsland"/>.
        /// </summary>
        /// <remarks>
        /// The island applies its default style when it is loaded by the owning XAML framework.
        /// </remarks>
        public DesktopFlyoutIsland()
        {
            DefaultStyleKey = typeof(DesktopFlyoutIsland);
            RegisterPropertyChangedCallback(VisibilityProperty, (s, e) => ((DesktopFlyoutIsland)s).OnIslandVisibilityChanged());
            LayoutUpdated += DesktopFlyoutIsland_LayoutUpdated;
            UpdateTemplateSettings();
        }

        /// <inheritdoc/>
        protected override void OnApplyTemplate()
        {
            base.OnApplyTemplate();

            Unloaded += DesktopFlyoutIsland_Unloaded;

            _propertyChangedCallbackTokenForCornerRadiusProperty = RegisterPropertyChangedCallback(CornerRadiusProperty, (s, e) => ((DesktopFlyoutIsland)s).OnCornerRadiusChanged());

            UpdateTemplateSettings();
        }

        internal void SetOwner(DesktopFlyout owner)
        {
            _owner = new(owner);

#if WASDK && !HAS_UNO
            UpdateOwnerBackdrop();
#endif
        }

        private static void OnIslandSizePropertyChanged(DependencyObject dependencyObject, DependencyPropertyChangedEventArgs e)
        {
            if (dependencyObject is not DesktopFlyoutIsland island)
                return;

            if (island._owner is not null && island._owner.TryGetTarget(out var owner))
                owner.OnIslandSizeChanged();
        }

        private void OnCornerRadiusChanged()
        {
            UpdateTemplateSettings();
        }

        private void OnIslandVisibilityChanged()
        {
            if (_owner is not null && _owner.TryGetTarget(out var owner))
                owner.OnIslandVisibilityChanged();
        }

        private void DesktopFlyoutIsland_LayoutUpdated(object? sender, object e)
        {
            var actualSize = new Size(ActualWidth, ActualHeight);
            var desiredSize = DesiredSize;
            if (_hasLayoutSnapshot &&
                AreClose(_lastActualSize, actualSize) &&
                AreClose(_lastDesiredSize, desiredSize))
            {
                return;
            }

            _hasLayoutSnapshot = true;
            _lastActualSize = actualSize;
            _lastDesiredSize = desiredSize;

            if (_owner is not null && _owner.TryGetTarget(out var owner))
                owner.OnIslandLayoutChanged();
        }

        private void UpdateTemplateSettings()
        {
#if WASDK && !HAS_UNO
            TemplateSettings.BackdropCornerRadius = new(
                GetBackdropCornerRadius(CornerRadius.TopLeft),
                GetBackdropCornerRadius(CornerRadius.TopRight),
                GetBackdropCornerRadius(CornerRadius.BottomRight),
                GetBackdropCornerRadius(CornerRadius.BottomLeft));
#endif
        }

#if WASDK && !HAS_UNO
        private static double GetBackdropCornerRadius(double cornerRadius)
        {
            return Math.Max(0D, cornerRadius > 0D ? cornerRadius - 1D : 0D);
        }

        internal void UpdateOwnerBackdrop()
        {
            TemplateSettings.SystemBackdrop = _owner is not null && _owner.TryGetTarget(out var owner)
                ? owner.CreateIslandSystemBackdrop()
                : null;
        }

        internal void ClearOwnerBackdrop()
        {
            TemplateSettings.SystemBackdrop = null;
        }
#endif

        private static bool AreClose(Size first, Size second)
        {
            return Math.Abs(first.Width - second.Width) < LayoutEpsilon &&
                Math.Abs(first.Height - second.Height) < LayoutEpsilon;
        }

        private void DesktopFlyoutIsland_Unloaded(object sender, RoutedEventArgs e)
        {
#if WASDK && !HAS_UNO
            ClearOwnerBackdrop();
#endif

            UnregisterPropertyChangedCallback(CornerRadiusProperty, _propertyChangedCallbackTokenForCornerRadiusProperty);

            Unloaded -= DesktopFlyoutIsland_Unloaded;
        }
    }
}
